### PR TITLE
New cluster join api

### LINKED ISSUE #4559: New cluster joining API
Right now joining a cluster requires the joining node to get pre-configured to have matching storage pools and networks before being allowed to join the cluster. This process is largely hidden to the user by the code we have in `lxd init`, but this leads to a suboptimal API for any external users. (@morphis)

Instead, what we should have is:
 - GET /1.0/cluster returning a list of node-specific configurations in a format that can be used by `lxd init` or other management tools
 - PUT 1/.0/cluster should accept a filled version of that same node-specific configuration alongside the cluster information

LXD will then pre-configure the joining node and go ahead with joining the cluster. If some change occurred in between, then the join will fail, we'll rollback the configuration on the joining node and the user can try again.

This will effectively get us down to just two API calls to join a cluster, one GET against the existing cluster and one PUT against the joining node.

### PR DESCRIPTION
**Pull request by freeekanayaka**
This extends the GET /1.0/cluster and PUT /1.0/cluster endpoints to provide information about the storage pools and networks required to join a cluster and automatically create them.

Closes #4559.

### CODE REVIEW

File: lxd/api_cluster.go [hunk 1]
```diff
@@ -46,14 +46,71 @@ func clusterGet(d *Daemon, r *http.Request) Response {
...
+// Fetch information about all node-specific configuration keys set on the
+// storage pools and networks of this cluster.
+func clusterGetMemberConfig(cluster *db.Cluster) ([]api.ClusterMemberConfigKey, error) {
+	var pools map[string]map[string]string
+	var networks map[string]map[string]string
+
+	keys := []api.ClusterMemberConfigKey{}
+
+	err := cluster.Transaction(func(tx *db.ClusterTx) error {
+		var err error
```
Comment (stgraber) [inline lxd/api_cluster.go]: line break

File: lxd/api_cluster.go [hunk 2]
```diff
@@ -46,14 +46,71 @@ func clusterGet(d *Daemon, r *http.Request) Response {
...
+		pools, err = tx.StoragePoolsNodeConfig()
+		if err != nil {
+			return errors.Wrapf(err, "failed to fetch storage pools configuration")
+		}
```
Comment (stgraber) [inline lxd/api_cluster.go]: line break

File: lxd/api_cluster.go [hunk 3]
```diff
@@ -46,14 +46,71 @@ func clusterGet(d *Daemon, r *http.Request) Response {
...
+		networks, err = tx.NetworksNodeConfig()
+		if err != nil {
+			return errors.Wrapf(err, "failed to fetch networks configuration")
+		}
```
Comment (stgraber) [inline lxd/api_cluster.go]: line break

File: lxd/api_cluster.go [hunk 4]
```diff
@@ -46,14 +46,71 @@ func clusterGet(d *Daemon, r *http.Request) Response {
...
+	var pools map[string]map[string]string
+	var networks map[string]map[string]string
+
+	keys := []api.ClusterMemberConfigKey{}
+
+	err := cluster.Transaction(func(tx *db.ClusterTx) error {
+		var err error
+		pools, err = tx.StoragePoolsNodeConfig()
+		if err != nil {
+			return errors.Wrapf(err, "failed to fetch storage pools configuration")
```
Comment (stgraber) [inline lxd/api_cluster.go]: Errors must start with a capital (and never end with a period)

File: lxd/api_cluster.go [hunk 5]
```diff
@@ -46,14 +46,71 @@ func clusterGet(d *Daemon, r *http.Request) Response {
...
+		}
+		networks, err = tx.NetworksNodeConfig()
+		if err != nil {
+			return errors.Wrapf(err, "failed to fetch networks configuration")
```
Comment (stgraber) [inline lxd/api_cluster.go]: Errors must start with a capital (and never end with a period)

File: lxd/api_cluster.go [hunk 6]
```diff
@@ -117,39 +174,43 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) Response {
...
-	}
-	for _, name := range poolNames {
-		_, pool, err := d.cluster.StoragePoolGet(name)
+		// The user has provided a server address, and no networking
+		// was setup on this node, let's do the job and open the port.
+		err := d.db.Transaction(func(tx *db.NodeTx) error {
+			config, err := node.ConfigLoad(tx)
+			if err != nil {
+				return errors.Wrap(err, "failed to load cluster config")
+			}
```
Comment (stgraber) [inline lxd/api_cluster.go]: line break

File: lxd/api_cluster.go [hunk 7]
```diff
@@ -117,39 +174,43 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) Response {
...
-		return SmartError(err)
-	}
-	for _, name := range poolNames {
-		_, pool, err := d.cluster.StoragePoolGet(name)
+		// The user has provided a server address, and no networking
+		// was setup on this node, let's do the job and open the port.
+		err := d.db.Transaction(func(tx *db.NodeTx) error {
+			config, err := node.ConfigLoad(tx)
+			if err != nil {
+				return errors.Wrap(err, "failed to load cluster config")
```
Comment (stgraber) [inline lxd/api_cluster.go]: Errors must start with a capital (and never end with a period)

File: lxd/api_cluster.go [hunk 8]
```diff
@@ -164,12 +225,69 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) Response {
...
-		// First request for this node to be added to the list of
-		// cluster nodes.
+
+		// If the user has provided a cluster password, setup the trust
+		// relationship by adding our own certificate to the cluster.
+		if req.ClusterPassword != "" {
+			err = cluster.SetupTrust(string(cert.PublicKey()), req.ClusterAddress,
+				string(req.ClusterCertificate), req.ClusterPassword)
+			if err != nil {
+				return errors.Wrap(err, "failed to setup cluster trust")
```
Comment (stgraber) [inline lxd/api_cluster.go]: Errors must start with a capital (and never end with a period)

File: lxd/api_cluster.go [hunk 9]
```diff
@@ -164,12 +225,69 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) Response {
...
+
+		// If node-specific configuration values are provided, it means
+		// the user wants to also initialize storage pools and
+		// networks.
+		if len(req.MemberConfig) > 0 {
+			// Connect to ourselves to initialize storage pools and
+			// networks using the API.
+			d, err := lxd.ConnectLXDUnix("", nil)
+			if err != nil {
+				return errors.Wrap(err, "failed to connect to local LXD")
```
Comment (stgraber) [inline lxd/api_cluster.go]: Errors must start with a capital (and never end with a period)

File: lxd/api_cluster.go [hunk 10]
```diff
@@ -164,12 +225,69 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) Response {
...
+			}
+
+			err = clusterInitMember(d, client, req.MemberConfig)
+			if err != nil {
+				return errors.Wrap(err, "failed to initialize member")
```
Comment (stgraber) [inline lxd/api_cluster.go]: Errors must start with a capital (and never end with a period)

File: lxd/api_cluster.go [hunk 11]
```diff
@@ -164,12 +225,69 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) Response {
...
+			}
+		}
+
+		// Get all defined storage pools and networks, so they can be compared
+		// to the ones in the cluster.
+		pools := []api.StoragePool{}
+		poolNames, err := d.cluster.StoragePools()
+		if err != nil && err != db.ErrNoSuchObject {
+			return err
+		}
```
Comment (stgraber) [inline lxd/api_cluster.go]: line break

File: lxd/api_cluster.go [hunk 12]
```diff
@@ -164,12 +225,69 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) Response {
...
+		for _, name := range poolNames {
+			_, pool, err := d.cluster.StoragePoolGet(name)
+			if err != nil {
+				return err
+			}
+			pools = append(pools, *pool)
+		}
```
Comment (stgraber) [inline lxd/api_cluster.go]: line break

File: lxd/api_cluster.go [hunk 13]
```diff
@@ -164,12 +225,69 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) Response {
...
+		pools := []api.StoragePool{}
+		poolNames, err := d.cluster.StoragePools()
+		if err != nil && err != db.ErrNoSuchObject {
+			return err
+		}
+		for _, name := range poolNames {
+			_, pool, err := d.cluster.StoragePoolGet(name)
+			if err != nil {
+				return err
+			}
```
Comment (stgraber) [inline lxd/api_cluster.go]: line break

File: lxd/api_cluster.go [hunk 14]
```diff
@@ -164,12 +225,69 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) Response {
...
+			pools = append(pools, *pool)
+		}
+		networks := []api.Network{}
+		networkNames, err := d.cluster.Networks()
+		if err != nil && err != db.ErrNoSuchObject {
+			return err
+		}
+		for _, name := range networkNames {
```
Comment (stgraber) [inline lxd/api_cluster.go]: line break

File: lxd/api_cluster.go [hunk 15]
```diff
@@ -164,12 +225,69 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) Response {
...
+			_, network, err := d.cluster.NetworkGet(name)
+			if err != nil {
+				return err
+			}
```
Comment (stgraber) [inline lxd/api_cluster.go]: line break

File: lxd/api_cluster.go [hunk 16]
```diff
@@ -164,12 +225,69 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) Response {
...
+			networks = append(networks, *network)
+		}
+
+		// Now request for this node to be added to the list of cluster
+		// nodes.
```
Comment (stgraber) [inline lxd/api_cluster.go]: needless wrapping

File: lxd/api_cluster.go [hunk 17]
```diff
@@ -329,6 +447,117 @@ func clusterPutDisable(d *Daemon) Response {
...
+//
+// We pass to LXD client instances, one connected to ourselves (the joining
+// node) and one connected to the target cluster node to join.
+func clusterInitMember(d, client lxd.ContainerServer, memberConfig []api.ClusterMemberConfigKey) error {
+	data := initDataNode{}
+
+	// Fetch all pools current defined in the cluster.
+	pools, err := client.GetStoragePools()
+	if err != nil {
+		return errors.Wrap(err, "failed to fetch information about cluster storage pools")
```
Comment (stgraber) [inline lxd/api_cluster.go]: Errors must start with a capital (and never end with a period)

File: lxd/api_cluster.go [hunk 18]
```diff
@@ -329,6 +447,117 @@ func clusterPutDisable(d *Daemon) Response {
...
+			continue
+		}
+
+		// Skip ceph pools since they have no node-specific key and
+		// don't need to be defined on joining nodes.
+		if pool.Driver == "ceph" {
+			continue
+		}
+
+		logger.Debugf("populating init data for storage pool %s", pool.Name)
```
Comment (stgraber) [inline lxd/api_cluster.go]: Log statements should follow same rules as errors

File: lxd/api_cluster.go [hunk 19]
```diff
@@ -329,6 +447,117 @@ func clusterPutDisable(d *Daemon) Response {
...
+			if config.Entity != "storage-pool" {
+				continue
+			}
+
+			if config.Name != pool.Name {
+				continue
+			}
+
+			if !shared.StringInSlice(config.Key, db.StoragePoolNodeConfigKeys) {
+				logger.Warnf("Ignoring config key %s for storage pool %s", config.Key, config.Name)
```
Comment (stgraber) [inline lxd/api_cluster.go]: Log statements should follow same rules as errors

File: lxd/api_cluster.go [hunk 20]
```diff
@@ -329,6 +447,117 @@ func clusterPutDisable(d *Daemon) Response {
...
+			post.Config[config.Key] = config.Value
+		}
+
+		data.StoragePools = append(data.StoragePools, post)
+	}
+
+	// Fetch all networks current defined in the cluster.
+	networks, err := client.GetNetworks()
+	if err != nil {
+		return errors.Wrap(err, "failed to fetch information about cluster networks")
```
Comment (stgraber) [inline lxd/api_cluster.go]: Errors must start with a capital (and never end with a period)

File: lxd/api_cluster.go [hunk 21]
```diff
@@ -329,6 +447,117 @@ func clusterPutDisable(d *Daemon) Response {
...
+			post.Config[config.Key] = config.Value
+		}
+
+		data.Networks = append(data.Networks, post)
+	}
+
+	revert, err := initDataNodeApply(d, data)
+	if err != nil {
+		revert()
+		return errors.Wrap(err, "failed to initialize storage pools and networks")
```
Comment (stgraber) [inline lxd/api_cluster.go]: Errors must start with a capital (and never end with a period)

File: lxd/api_cluster.go [hunk 22]
```diff
@@ -329,6 +447,117 @@ func clusterPutDisable(d *Daemon) Response {
...
 }
 
+// Initialize storage pools and networks on this node.
+//
+// We pass to LXD client instances, one connected to ourselves (the joining
+// node) and one connected to the target cluster node to join.
+func clusterInitMember(d, client lxd.ContainerServer, memberConfig []api.ClusterMemberConfigKey) error {
+	data := initDataNode{}
+
+	// Fetch all pools current defined in the cluster.
```
Comment (stgraber) [inline lxd/api_cluster.go]: currently

File: lxd/api_cluster.go [hunk 23]
```diff
@@ -329,6 +447,117 @@ func clusterPutDisable(d *Daemon) Response {
...
+				continue
+			}
+
+			post.Config[config.Key] = config.Value
+		}
+
+		data.StoragePools = append(data.StoragePools, post)
+	}
+
+	// Fetch all networks current defined in the cluster.
```
Comment (stgraber) [inline lxd/api_cluster.go]: currently

File: lxd/cluster/connect.go [hunk 24]
```diff
@@ -112,7 +112,7 @@ func SetupTrust(cert, targetAddress, targetCert, targetPassword string) error {
 	post.Type = "client"
 	err = target.CreateCertificate(post)
 	if err != nil && err.Error() != "Certificate already in trust store" {
-		return errors.Wrap(err, "failed client cert to cluster")
+		return errors.Wrap(err, "failed to add client cert to cluster")
```
Comment (stgraber) [inline lxd/cluster/connect.go]: Errors must start with a capital (and never end with a period)

Comment (stgraber) [reply lxd/cluster/connect.go]: I realize this one was already there, but might as well fix it since you're changing it :)

File: test/includes/lxd.sh [hunk 25]
```diff
@@ -31,8 +31,8 @@ spawn_lxd() {
         cp deps/server.crt "${lxddir}"
         cp deps/server.key "${lxddir}"
     else
-        cp deps/server-alt.crt "${lxddir}"
-        cp deps/server-alt.key "${lxddir}"
+        cp deps/server-alt.crt "${lxddir}"/server.crt
+        cp deps/server-alt.key "${lxddir}"/server.key
```
Comment (stgraber) [inline test/includes/lxd.sh]: Good catch, can you split into its own commit please?

File: lxd/api_cluster.go (hunk 1 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 2 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 3 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 5 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 7 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 8 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 9 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 10 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 11 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 13 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 15 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 16 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 17 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 19 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 21 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 22 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/api_cluster.go (hunk 23 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: lxd/cluster/connect.go (hunk 24 above)
Comment (freeekanayaka) [reply lxd/cluster/connect.go]: Fixed.

File: test/includes/lxd.sh (hunk 25 above)
Comment (freeekanayaka) [reply test/includes/lxd.sh]: Done.

File: lxd/api_cluster.go (hunk 6 above)
Comment (freeekanayaka) [reply lxd/api_cluster.go]: Fixed.

File: doc/rest-api.md [hunk 26]
```diff
@@ -2607,9 +2621,24 @@ Input (request to join an existing cluster):
 
     {
         "server_name": "node2",
+        "server_address": "10.1.1.102:8443",
         "enabled": true,
         "cluster_address": "10.1.1.101:8443",
         "cluster_certificate": "-----BEGIN CERTIFICATE-----MIFf\n-----END CERTIFICATE-----",
+		"cluster_password": "sekret",
```
Comment (stgraber) [inline doc/rest-api.md]: Indentation looks wrong here

**Comment by stgraber** [comment]
@freeekanayaka looks fine except for the broken indent in the docs.

Once you've got that fixed, can you prepare a branch which has `lxd init` itself ported over to using the new API? That should mostly affect the interactive path, effectively making sure that the only API calls for an interactive `lxd init` are:

 - local: GET /1.0 (init local client)
 - cluster: GET /1.0 (init cluster client)
 - cluster: POST /1.0/certificates (add client cert if not already trusted)
 - cluster: GET /1.0/cluster (retrieve the questions)
 - local: PUT /1.0/cluster (join the cluster)

File: doc/rest-api.md (hunk 26 above)
Comment (freeekanayaka) [reply doc/rest-api.md]: Fixed.
